### PR TITLE
feat: mark finalized blocks

### LINKED ISSUE #1036: [Task]: Subscribe to `commitment-info/finalized` topic to mark finalized blocks
### Task description

All accepted or confirmed blocks are finalized once the corresponding slot is finalized, so once you get a finalized commitment you can automatically mark all blocks of that slot that you had accepted/confirmed as finalized

- the slot is IN the block ID
- when the slot gets finalized and the block was accepted/approved, the block is finalized

### PR DESCRIPTION
**Pull request by VmMad**
# Description of change

Fixes #1036 
Depends on [iota-sdk/1962](https://github.com/iotaledger/iota-sdk/pull/1962)

## Type of change

Choose a type of change, and delete any options that are not relevant.

- Enhancement (a non-breaking change which adds functionality)

## How the change has been tested

1. Change the finalized color to yellow `#FFFF00` 
2. Wait until there are ~400 blocks in the screen and pause the visualizer.
3. Wait a moment until the blocks in the screen are marked as finalized
 
## Change checklist

Add an `x` to the boxes that are relevant to your changes, and delete any items that are not.

- [x] My code follows the contribution guidelines for this project
- [x] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: client/src/features/visualizer-threejs/VisualizerInstance.tsx [hunk 1]
```diff
@@ -217,9 +222,37 @@ const VisualizerInstance: React.FC<RouteComponentProps<VisualizerRouteProps>> =
...
+                if (selectedColor) {
+                    addToColorQueue(blockId, selectedColor);
+                }
+            });
+        }
+
+        removeConfirmedBlocksSlot(slot);
+    }
+
+    function getSlotIndexFromBlockId(blockId: BlockId): SlotIndex {
```
Comment (VmMad) [inline client/src/features/visualizer-threejs/VisualizerInstance.tsx:250]: will remove this function when the iota-sdk PR gets merged

File: client/src/features/visualizer-threejs/store/tangle.ts [hunk 2]
```diff
@@ -202,5 +209,33 @@ export const useTangleStore = create<TangleState>()(
...
+        removeConfirmedBlocksSlot: (slot) => {
+            set((state) => {
+                state.confirmedBlocksBySlot.delete(slot);
+
+                // Cleanup all slots that are lower than the current slot
+                for (const existingSlot of state.confirmedBlocksBySlot.keys()) {
+                    if (existingSlot < slot) {
+                        state.confirmedBlocksBySlot.delete(existingSlot);
+                    }
+                }
```
Comment (VmMad) [inline client/src/features/visualizer-threejs/store/tangle.ts:232]: Question for reviewers: 
We can safely remove previous slots when one gets confirmed, right? Like if 592 gets confirmed, we can remove 591, 590...?

Comment (msarcev) [reply client/src/features/visualizer-threejs/store/tangle.ts:234]: What does "removing a slot" in this context do ? Do we use that info only to colour ?

File: client/src/features/visualizer-threejs/VisualizerInstance.tsx [hunk 3]
```diff
@@ -167,11 +171,22 @@ const VisualizerInstance: React.FC<RouteComponentProps<VisualizerRouteProps>> =
         };
     }, [runListeners]);
 
+    useEffect(() => {
```
Comment (msarcev) [inline client/src/features/visualizer-threejs/VisualizerInstance.tsx:174]: Why was this added ? There an identical `useEffect` just above this `useEffect`

File: client/src/features/visualizer-threejs/store/tangle.ts (hunk 2 above)
Comment (VmMad) [reply client/src/features/visualizer-threejs/store/tangle.ts:234]: in this map yes, its like a queue to color the blocks.
I wanted to make a "fallback cleanup" in case something fails we don't store unnecessary stuff

Comment (VmMad) [reply client/src/features/visualizer-threejs/store/tangle.ts:234]: Perhaps I need to rename the map 🤔

File: client/src/features/visualizer-threejs/VisualizerInstance.tsx (hunk 3 above)
Comment (VmMad) [reply client/src/features/visualizer-threejs/VisualizerInstance.tsx:174]: Merge debris :(

File: client/src/features/visualizer-threejs/store/tangle.ts (hunk 2 above)
Comment (msarcev) [reply client/src/features/visualizer-threejs/store/tangle.ts:234]: I guess we can safely remove then, if the old one are "used" and removing won't affect the colours.

Comment (VmMad) [reply client/src/features/visualizer-threejs/store/tangle.ts:234]: @panteleymonchuk has this happened to you? From what I've read, when a slot is finalized and the blocks are unconfirmed, they become orphans. And since slot 592 should be finished, that means that the block in the slot 590 was removed from the consensus

Comment (VmMad) [reply client/src/features/visualizer-threejs/store/tangle.ts:234]: You mean that the operation for getting the confirmation for the finalized block on the slot 590 gets "delayed"? In that case, we should have set the blocks for 590 as finalized already. So, if it's then added to the slot 590 for finalized blocks, if the slot 593 finishes, it will iterate over the map and remove the slots with less index than 593 and remove them, since it should have already finished. I don't think it makes sense to paint a block later when the slot has already finished

**Review by msarcev — APPROVED** [review]

File: api/src/services/nova/feed/novaFeed.ts [hunk 4]
```diff
@@ -112,6 +112,22 @@ export class NovaFeed {
...
+                const deserializedMessage: { topic: string; payload: string } = JSON.parse(message);
+                const slotCommitment: SlotCommitment = JSON.parse(deserializedMessage.payload);
+                const slotFinalized = slotCommitment.slot;
+
+                const update: Partial<IFeedUpdate> = { slotFinalized };
+
+                // eslint-disable-next-line no-void
+                void this.broadcastBlock(update);
+            } catch {
+                logger.error("[NovaFeed]: Failed broadcasting block-metadata downstream.");
```
Comment (begonaalvarezd) [inline api/src/services/nova/feed/novaFeed.ts:128]: this error seems copied from somewhere else

File: client/src/features/visualizer-threejs/VisualizerInstance.tsx [hunk 5]
```diff
@@ -67,6 +68,9 @@ const VisualizerInstance: React.FC<RouteComponentProps<VisualizerRouteProps>> =
     const blockMetadata = useTangleStore((s) => s.blockMetadata);
     const indexToBlockId = useTangleStore((s) => s.indexToBlockId);
     const clickedInstanceId = useTangleStore((s) => s.clickedInstanceId);
+    const confirmedBlocksBySlot = useTangleStore((s) => s.confirmedBlocksBySlot);
```
Comment (begonaalvarezd) [inline client/src/features/visualizer-threejs/VisualizerInstance.tsx:71]: add a comment explaining this covers confirmed and accepted blocks

File: client/src/features/visualizer-threejs/store/tangle.ts [hunk 6]
```diff
@@ -64,6 +65,11 @@ interface TangleState {
 
     updateBlockIdToAnimationPosition: (updatedPositions: Map<string, IBlockInitPosition>) => void;
     resetConfigState: () => void;
+
+    // Handle finalized blocks
```
Comment (begonaalvarezd) [inline client/src/features/visualizer-threejs/store/tangle.ts:69]: this comment is misleading because these are for accepted/confirmed blocks, not finalized

**Review by begonaalvarezd — CHANGES_REQUESTED** [review]

**Review by begonaalvarezd — APPROVED** [review]
